server/reaction_hooks.go

```diff
@@ -0,0 +1,108 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+)
+
+// ReactionHasBeenAdded is invoked after the reaction has been committed to the database.
+//
+// Note that this method will be called for reactions added by plugins, including the plugin that
+// added the reaction.
+//
+// This demo implementation logs a message to the demo channel whenever a reaction is added to a post.
+func (p *Plugin) ReactionHasBeenAdded(c *plugin.Context, reaction *model.Reaction) {
+	configuration := p.getConfiguration()
+
+	if configuration.disabled {
+		return
+	}
+
+	user, err := p.API.GetUser(reaction.UserId)
+	if err != nil {
+		p.API.LogError("failed to query user", "user_id", reaction.UserId)
+		return
+	}
+
+	post, err := p.API.GetPost(reaction.PostId)
+	if err != nil {
+		p.API.LogError("failed to query post", "post_id", reaction.PostId)
+		return
+	}
+
+	channel, err := p.API.GetChannel(post.ChannelId)
+	if err != nil {
+		p.API.LogError("failed to query channel", "channel_id", post.ChannelId)
+		return
+	}
+
+	team, err := p.API.GetTeam(channel.TeamId)
+	if err != nil {
+		p.API.LogError("failed to query team", "team_id", channel.TeamId)
+		return
+	}
```

Comment:
If the emojis is removing from a post in a DM, this call will fail. You can use construct the link using `_redirect/pl/$POST_ID` similar to https://github.com/matterpoll/matterpoll/blob/45f095875a98fb1d4f3f166851c86f41b987493e/server/plugin/api.go#L665

Reply:
@hanzei - Ah, great insight, thank you! Didn't realize that as another way to construct the link.

One thought though... I just tested a direct message, and the code actually fails before ever getting to this point.  Specifically the call before it to `GetChannel(post.ChannelId)`  fails for a direct message, so the demo plugin never gets to log the event.  I checked and it looks like this is also happening in the current setup for `MessageHasBeenPosted` (e.g. if you post the plugin secret in a direct message, it will fail and not log to the demo channel)... Potentially other places too?

As written, these are two sets of functionality (in the context of this plugin) that fail during direct messages.  Do we want to figure out a way to adapt them to work for DM's (if so, I can change the URL above)... Or do we want to leave them as-is, knowing they will fail for DM's?

Reply:
Are you sure? For me it fails with `failed to query team` while the channel gets fetched correctly. This is also the behaviour that I was expecting.

Reply:
Whoops, you're correct, I got ahead of myself and tested after I changed the code.  However, the overall hook still fails to post the message to the demo channel, since there's no team to correlate it to.

I went ahead and updated the URL to the pattern you mentioned.

We're still left with `postPluginMessage()` failing when called for a post that was from a DM (so you'll never see the logged message for a DM).  Do you want to accept that as-designed (it currently impacts the message hooks too)... or do you want to consider options for how to keep the functionality when executed from a DM?  For example, when the correct demo channel cannot be found, we could always use the bot to direct message the log message back to the initiator...